### PR TITLE
Receive chat messages via notification websocket

### PR DESCRIPTION
**Pull request by smoogipoo**
Prereqs:
- [x] https://github.com/ppy/osu/pull/20990
- [x] Need to test tournament (should work, in theory).

I've implemented two connectors - a websocket version which is new interesting one, and a polling version that mimics the old functionality by polling every 1 second and is useful for tests.

### CODE REVIEW

File: osu.Game/Online/SocketClientConnector.cs
Comment (peppy) [inline osu.Game/Online/SocketClientConnector.cs:68]: actual test failure?

Comment (smoogipoo) [reply osu.Game/Online/SocketClientConnector.cs:68]: Yes, it will be. I'll take a look.

Comment (smoogipoo) [reply osu.Game/Online/SocketClientConnector.cs:71]: Well. I couldn't repro the test failures, but I did some sanitisation and ran CI a few times and it seems fine...

File: osu.Game/Online/Notifications/NotificationsClientConnector.cs
Comment (peppy) [inline osu.Game/Online/Notifications/NotificationsClientConnector.cs:29]: I don't understand why this is a thing. Is there a case we *wouldn't* enable chat?

File: osu.Game/Online/Notifications/NotificationsClient.cs
Comment (peppy) [inline osu.Game/Online/Notifications/NotificationsClient.cs:34]: I doubt this is ever going to be a thing, but setting this to false doesn't seem to do anything.

Also curious why this is a `bool` rather than a `StartChat` method like the invocation.

File: osu.Game/Online/Chat/ChannelManager.cs
Comment (peppy) [inline osu.Game/Online/Chat/ChannelManager.cs:77]: Kinda weird passing the new parameter here even though it can be fetched directly from `IAPIProvider`. Was there a reason behind this?

File: osu.Game/Online/Chat/ChannelManager.cs
Comment (peppy) [inline osu.Game/Online/Chat/ChannelManager.cs:109]: Rather than doing it this way, I think it would be safer to only queue the next request when the previous one completes.

ie. change this to `Scheduler.Add(performChatAckRequest)` and schedule the next call within that new method on `Success` or `Fail`.

File: osu.Game/Online/Notifications/Polling/PollingNotificationsClient.cs
Comment (peppy) [inline osu.Game/Online/Notifications/Polling/PollingNotificationsClient.cs:13]: I'd consider moving this to the test namespace (along with `PollingNotificationsClientConnector`) since it's only used there.

**Comment by peppy** [comment]
Doesn't seem to work for me on production:

```csharp
[network] 2022-11-02 08:37:31 [verbose]: Failing request osu.Game.Online.API.Requests.GetNotificationsRequest (Newtonsoft.Json.JsonReaderException: Unexpected character encountered while parsing value: {. Path 'notifications[7].details.reply_to', line 1, position 2315.
[network] 2022-11-02 08:37:31 [verbose]: at Newtonsoft.Json.JsonTextReader.ReadStringValue(ReadType readType)
[network] 2022-11-02 08:37:31 [verbose]: at Newtonsoft.Json.JsonTextReader.ReadAsString()
[network] 2022-11-02 08:37:31 [verbose]: at Newtonsoft.Json.JsonReader.ReadForType(JsonContract contract, Boolean hasConverter)
[network] 2022-11-02 08:37:31 [verbose]: at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.PopulateDictionary(IDictionary dictionary, JsonReader reader, JsonDictionaryContract contract, JsonProperty containerProperty, String id)
[network] 2022-11-02 08:37:31 [verbose]: at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateObject(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
[network] 2022-11-02 08:37:31 [verbose]: at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.CreateValueInternal(JsonReader reader, Type objectType, JsonContract contract, JsonProperty member, JsonContainerContract containerContract, JsonProperty containerMember, Object existingValue)
[network] 2022-11-02 08:37:31 [verbose]: at Newtonsoft.Json.Serialization.JsonSerializerInternalReader.SetPropertyValue(JsonProperty property, JsonConverter propertyConverter, JsonContainerContract containerContract, JsonProperty containerProperty, JsonReader reader, Object target)
```

File: osu.Game/Online/Chat/ChannelManager.cs
Comment (smoogipoo) [reply osu.Game/Online/Chat/ChannelManager.cs:77]: It's uh... legacy code :smile: . It can be constructed inside this class now.

**Comment by peppy** [comment]
Above issue with no chat displaying was fixed via 4f8e912.

**Comment by smoogipoo** [comment]
Blocking this while looking into the local echo issue brought up IRL (messages sent by the local user through osu!web don't appear ingame).

File: osu.Game/Online/Chat/ChannelManager.cs
Comment (smoogipoo) [reply osu.Game/Online/Chat/ChannelManager.cs]: I'm not sure it's safer going through events, but sure. I've made the change to the best of my ability in making it safe.

**Comment by smoogipoo** [comment]
I believe I've fixed most of the issues I've come across, but there's some weirdness with osu-web, possibly related to https://github.com/ppy/osu-notification-server/pull/47 as the symptoms I'm showing are similar - if osu!web and the client are both in chat the client stops receiving messages.

Will wait for the PR above to be merged/deployed before I unblock this/test further.

**Review by peppy — APPROVED** [review]

**Comment by adanraj** [comment]
comment g osu lazer

**Pull request merged**
